File: bukkit/elysium-characters-bukkit/src/main/java/com/seventh_root/elysium/characters/bukkit/character/BukkitCharacterProvider.java
package com.seventh_root.elysium.characters.bukkit.character;

import com.seventh_root.elysium.api.character.CharacterProvider;
import com.seventh_root.elysium.api.player.ElysiumPlayer;
import com.seventh_root.elysium.characters.bukkit.ElysiumCharactersBukkit;
import com.seventh_root.elysium.players.bukkit.BukkitPlayer;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class BukkitCharacterProvider implements CharacterProvider<BukkitCharacter> {

    private final ElysiumCharactersBukkit plugin;

    public BukkitCharacterProvider(ElysiumCharactersBukkit plugin) {
        this.plugin = plugin;
    }

    @Override
    public BukkitCharacter getCharacter(int id) {
        return plugin.getCore().getDatabase().getTable(BukkitCharacter.class).get(id);
    }

    @Override
    public BukkitCharacter getActiveCharacter(ElysiumPlayer player) {
        try (
                Connection connection = plugin.getCore().getDatabase().createConnection();
                PreparedStatement statement = connection.prepareStatement(
                        "SELECT character_id FROM player_character WHERE player_id = ?"
                )
        ) {
            statement.setInt(1, player.getId());
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                return getCharacter(resultSet.getInt("character_id"));
            }
        } catch (SQLException exception) {
            exception.printStackTrace();
        }
        return null;
    }

    @Override
    public void setActiveCharacter(ElysiumPlayer player, BukkitCharacter character) {
        BukkitCharacter oldCharacter = getActiveCharacter(player);
        if (oldCharacter != null) {
            if (player instanceof BukkitPlayer) {
                OfflinePlayer offlineBukkitPlayer = ((BukkitPlayer) player).getBukkitPlayer();
                if (offlineBukkitPlayer.isOnline()) {
                    Player bukkitPlayer = offlineBukkitPlayer.getPlayer();
                    oldCharacter.setHelmet(bukkitPlayer.getInventory().getHelmet());
                    oldCharacter.setChestplate(bukkitPlayer.getInventory().getChestplate());
                    oldCharacter.setLeggings(bukkitPlayer.getInventory().getLeggings());
                    oldCharacter.setBoots(bukkitPlayer.getInventory().getBoots());
                    oldCharacter.setInventoryContents(bukkitPlayer.getInventory().getContents());
                    oldCharacter.setLocation(bukkitPlayer.getLocation());
                    oldCharacter.setHealth(bukkitPlayer.getHealth());
                    oldCharacter.setFoodLevel(bukkitPlayer.getFoodLevel());
                    updateCharacter(oldCharacter);
                }
            }
        }
        if (character != null) {
            try (
                    Connection connection = plugin.getCore().getDatabase().createConnection();
                    PreparedStatement statement = connection.prepareStatement(
                            "INSERT INTO player_character(player_id, character_id) VALUES(?, ?) ON DUPLICATE KEY UPDATE character_id = VALUES(character_id)"
                    )
            ) {
                statement.setInt(1, player.getId());
                statement.setInt(2, character.getId());
                statement.executeUpdate();
            } catch (SQLException exception) {
                exception.printStackTrace();
            }
            if (player instanceof BukkitPlayer) {
                OfflinePlayer offlineBukkitPlayer = ((BukkitPlayer) player).getBukkitPlayer();
                if (offlineBukkitPlayer.isOnline()) {
                    Player bukkitPlayer = offlineBukkitPlayer.getPlayer();
                    bukkitPlayer.getInventory().setHelmet(character.getHelmet());
                    bukkitPlayer.getInventory().setChestplate(character.getChestplate());
                    bukkitPlayer.getInventory().setLeggings(character.getLeggings());
                    bukkitPlayer.getInventory().setBoots(character.getBoots());
                    bukkitPlayer.getInventory().setContents(character.getInventoryContents());
                    bukkitPlayer.teleport(character.getLocation());
                    bukkitPlayer.setHealth(character.getHealth());
                    bukkitPlayer.setMaxHealth(character.getMaxHealth());
                    bukkitPlayer.setFoodLevel(character.getFoodLevel());
                }
            }
        } else if (oldCharacter != null) {
            try (
                    Connection connection = plugin.getCore().getDatabase().createConnection();
                    PreparedStatement statement = connection.prepareStatement(
                            "DELETE FROM player_character WHERE player_id = ? AND character_id = ?"
                    )
            ) {
                statement.setInt(1, player.getId());
                statement.setInt(2, oldCharacter.getId());
                statement.executeUpdate();
            } catch (SQLException exception) {
                exception.printStackTrace();
            }
        }
    }

    @Override
    public Collection<? extends BukkitCharacter> getCharacters(ElysiumPlayer player) {
        try (
                Connection connection = plugin.getCore().getDatabase().createConnection();
                PreparedStatement statement = connection.prepareStatement(
                        "SELECT id FROM bukkit_character WHERE player_id = ? ORDER BY id"
                )
        ) {
            statement.setInt(1, player.getId());
            ResultSet resultSet = statement.executeQuery();
            List<BukkitCharacter> characters = new ArrayList<>();
            while (resultSet.next()) {
                characters.add(getCharacter(resultSet.getInt("id")));
            }
            return characters;
        } catch (SQLException exception) {
            exception.printStackTrace();
        }
        return null;
    }

    @Override
    public int addCharacter(BukkitCharacter character) {
        return plugin.getCore().getDatabase().getTable(BukkitCharacter.class).insert(character);
    }

    @Override
    public void removeCharacter(BukkitCharacter character) {
        setActiveCharacter(character.getPlayer(), null);
        plugin.getCore().getDatabase().getTable(BukkitCharacter.class).delete(character);
    }

    @Override
    public void updateCharacter(BukkitCharacter character) {
        if (plugin.getConfig().getBoolean("characters.delete-character-on-death") && character.isDead()) {
            removeCharacter(character);
        } else {
            plugin.getCore().getDatabase().getTable(BukkitCharacter.class).update(character);
        }
    }

}

File: bukkit/elysium-characters-bukkit/src/main/kotlin/com/seventh_root/elysium/characters/bukkit/character/BukkitCharacterProvider.kt
package com.seventh_root.elysium.characters.bukkit.character

import com.seventh_root.elysium.api.character.CharacterProvider
import com.seventh_root.elysium.api.player.ElysiumPlayer
import com.seventh_root.elysium.characters.bukkit.ElysiumCharactersBukkit
import com.seventh_root.elysium.core.database.use
import com.seventh_root.elysium.players.bukkit.BukkitPlayer
import org.ehcache.Cache
import org.ehcache.CacheManager
import org.ehcache.config.builders.CacheConfigurationBuilder
import org.ehcache.config.builders.CacheManagerBuilder
import java.sql.SQLException
import java.util.*

class BukkitCharacterProvider : CharacterProvider<BukkitCharacter> {

    private val plugin: ElysiumCharactersBukkit
    private val activeCharacterCacheManager: CacheManager
    private val activeCharacterCache: Cache<Int, Int>

    constructor(plugin: ElysiumCharactersBukkit) {
        this.plugin = plugin
        this.activeCharacterCacheManager = CacheManagerBuilder.newCacheManagerBuilder().build(true)
        activeCharacterCache = activeCharacterCacheManager.createCache("cache", CacheConfigurationBuilder.newCacheConfigurationBuilder(Int::class.javaObjectType, Int::class.javaObjectType).build())
    }

    override fun getCharacter(id: Int): BukkitCharacter? {
        return plugin.core!!.database.getTable(BukkitCharacter::class.java)!![id]
    }

    override fun getActiveCharacter(player: ElysiumPlayer): BukkitCharacter? {
        val playerId = player.id
        if (activeCharacterCache.containsKey(playerId)) {
            return getCharacter(activeCharacterCache.get(playerId) as Int)
        } else {
            try {
                var character: BukkitCharacter? = null
                plugin.core!!.database.createConnection().use { connection ->
                    connection.prepareStatement(
                            "SELECT character_id FROM player_character WHERE player_id = ?").use({ statement ->
                        statement.setInt(1, player.id)
                        val resultSet = statement.executeQuery()
                        if (resultSet.next()) {
                            val characterId = resultSet.getInt("character_id")
                            character = getCharacter(characterId)
                            activeCharacterCache.put(playerId, characterId)
                        }
                    })
                }
                return character
            } catch (exception: SQLException) {
                exception.printStackTrace()
            }
        }
        return null
    }

    override fun setActiveCharacter(player: ElysiumPlayer, character: BukkitCharacter?) {
        val oldCharacter = getActiveCharacter(player)
        if (oldCharacter != null) {
            if (player is BukkitPlayer) {
                val offlineBukkitPlayer = player.bukkitPlayer
                if (offlineBukkitPlayer.isOnline) {
                    val bukkitPlayer = offlineBukkitPlayer.player
                    oldCharacter.inventoryContents = bukkitPlayer.inventory.contents
                    oldCharacter.helmet = bukkitPlayer.inventory.helmet
                    oldCharacter.chestplate = bukkitPlayer.inventory.chestplate
                    oldCharacter.leggings = bukkitPlayer.inventory.leggings
                    oldCharacter.boots = bukkitPlayer.inventory.boots
                    oldCharacter.location = bukkitPlayer.location
                    oldCharacter.health = bukkitPlayer.health
                    oldCharacter.foodLevel = bukkitPlayer.foodLevel
                    updateCharacter(oldCharacter)
                }
            }
        }
        if (character != null) {
            try {
                plugin.core!!.database.createConnection().use { connection ->
                    connection.prepareStatement(
                            "INSERT INTO player_character(player_id, character_id) VALUES(?, ?) ON DUPLICATE KEY UPDATE character_id = VALUES(character_id)").use({ statement ->
                        statement.setInt(1, player.id)
                        statement.setInt(2, character.id)
                        statement.executeUpdate()
                    })
                }
            } catch (exception: SQLException) {
                exception.printStackTrace()
            }

            if (player is BukkitPlayer) {
                val offlineBukkitPlayer = player.bukkitPlayer
                if (offlineBukkitPlayer.isOnline) {
                    val bukkitPlayer = offlineBukkitPlayer.player
                    bukkitPlayer.inventory.contents = character.inventoryContents
                    bukkitPlayer.inventory.helmet = character.helmet
                    bukkitPlayer.inventory.chestplate = character.chestplate
                    bukkitPlayer.inventory.leggings = character.leggings
                    bukkitPlayer.inventory.boots = character.boots
                    bukkitPlayer.teleport(character.location)
                    bukkitPlayer.maxHealth = character.maxHealth
                    bukkitPlayer.health = character.health
                    bukkitPlayer.foodLevel = character.foodLevel
                }
            }
            activeCharacterCache.put(player.id, character.id)
        } else if (oldCharacter != null) {
            try {
                plugin.core!!.database.createConnection().use { connection ->
                    connection.prepareStatement(
                            "DELETE FROM player_character WHERE player_id = ? AND character_id = ?").use({ statement ->
                        statement.setInt(1, player.id)
                        statement.setInt(2, oldCharacter.id)
                        statement.executeUpdate()
                    })
                }
            } catch (exception: SQLException) {
                exception.printStackTrace()
            }
            activeCharacterCache.remove(player.id)
        }
    }

    override fun getCharacters(player: ElysiumPlayer): Collection<BukkitCharacter> {
        try {
            val characters: MutableList<BukkitCharacter> = ArrayList()
            plugin.core!!.database.createConnection().use { connection ->
                connection.prepareStatement(
                        "SELECT id FROM bukkit_character WHERE player_id = ? ORDER BY id").use({ statement ->
                    statement.setInt(1, player.id)
                    val resultSet = statement.executeQuery()
                    while (resultSet.next()) {
                        characters.add(getCharacter(resultSet.getInt("id"))!!)
                    }

                })
            }
            return characters
        } catch (exception: SQLException) {
            exception.printStackTrace()
        }
        return emptyList()
    }

    override fun addCharacter(character: BukkitCharacter): Int {
        return plugin.core!!.database.getTable(BukkitCharacter::class.java)!!.insert(character)
    }

    override fun removeCharacter(character: BukkitCharacter) {
        val player = character.player
        if (player != null)
            setActiveCharacter(player, null)
        plugin.core!!.database.getTable(BukkitCharacter::class.java)!!.delete(character)
    }

    override fun updateCharacter(character: BukkitCharacter) {
        if (plugin.config.getBoolean("characters.delete-character-on-death") && character.isDead) {
            removeCharacter(character)
        } else {
            plugin.core!!.database.getTable(BukkitCharacter::class.java)!!.update(character)
        }
    }

}
